### PR TITLE
Remove Event.category_chain and enforce consistency

### PR DESCRIPTION
**Pull request by ThiefMaster**
`indico db downgrade 535f817f2533` beforehand

Reasons for removing the column:

- It is impossible to keep it fully in sync (move category in
  transaction 1, create event in that category while the move
  transaction is running)
- In almost all cases it is actually **slower** to query events based
  on the category_chain column compared to using the CTE
- The consistency checks (besides not covering all cases, see point 1)
  were extremely expensive for very big categories (up to ~20s)
- Synchronizing the category chains for big categories after a move was
  also very expensive (>1m

Consistency checks:

- nothing in a deleted category that's not deleted
- no cycles in the category chain

### CODE REVIEW

File: indico/modules/categories/models/categories.py [hunk 1]
```diff
@@ -256,6 +256,21 @@ def can_create_events(self, user):
         """Check whether the user can create events in the category."""
         return user and (not self.event_creation_restricted or self.can_manage(user, role='create'))
 
+    def move(self, target):
+        """Move the category into another category."""
+        assert not self.is_root
```
Comment (OmeGak) [inline indico/modules/categories/models/categories.py]: Do we check a category not being able to be moved to any of the subcategories? Shouldn't this check belong there or the other here?

Comment (ThiefMaster) [inline indico/modules/categories/models/categories.py]: You mean getting a cycle between categories? That's checked using a constraint trigger. I'd rather have a user-friendly check for this in the RH (or even better, disallow it on the widget level when selecting the move target)

File: indico/modules/categories/models/categories.py [hunk 2]
```diff
@@ -256,6 +256,50 @@ def can_create_events(self, user):
...
+                     .cte(recursive=True))
+        rec_query = (select([cat_alias.id,
+                             cte_query.c.path.op('||')(path_column),
+                             cte_query.c.is_deleted | cat_alias.is_deleted])
+                     .where(cat_alias.parent_id == cte_query.c.id))
+        cte_query = cte_query.union_all(rec_query)
+        return cte_query
+
+    @property
+    def deep_children_query(self):
```
Comment (OmeGak) [inline indico/modules/categories/models/categories.py]: Why not `deep_subcategory_query`? Questionable naming as @pferreir mentioned.

Comment (ThiefMaster) [inline indico/modules/categories/models/categories.py]: Not part of the PR ;) I just moved it. But yes, can be changed
